exchange/exchange.go

```diff
@@ -0,0 +1,221 @@
+			// Only add non-null seat bids
+			// Possible performance improvement by grabbing a pointer to the current seatBid element, passing it to
+			// MakeSeatBid, and then building the seatBid in place, rather than copying. Probably more confusing than
+			// its worth
+			sb := e.MakeSeatBid(adapterBids[a], a, adapterExtra)
+			seatBids = append(seatBids, *sb)
+		}
+	}
+	bidResponse.SeatBid = seatBids
+
+	bidResponseExt := e.MakeExtBidResponse(adapterBids, adapterExtra, bidRequest.Test, errList)
+	ext, err := json.Marshal(bidResponseExt)
+	bidResponse.Ext = ext
+
+	return bidResponse, err
+}
+
+// Extract all the data from the SeatBids and build the ExtBidResponse
+func (e *exchange) MakeExtBidResponse(adapterBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid, adapterExtra map[openrtb_ext.BidderName]*seatResponseExtra, test int8, errList []error) *openrtb_ext.ExtBidResponse {
+	bidResponseExt := &openrtb_ext.ExtBidResponse{
+		Errors: make(map[openrtb_ext.BidderName][]string, len(adapterBids)),
+		ResponseTimeMillis: make(map[openrtb_ext.BidderName]int, len(adapterBids)),
+	}
+	if test == 1 {
+		bidResponseExt.Debug = &openrtb_ext.ExtResponseDebug{
+			HttpCalls: make(map[openrtb_ext.BidderName][]*openrtb_ext.ExtHttpCall),
+		}
+	}
+
+	for a, b := range adapterBids {
+		if b != nil {
+			if test == 1 {
+				// Fill debug info
+				bidResponseExt.Debug.HttpCalls[a] = b.httpCalls
+			}
+		}
+		// Only make an entry for bidder errors if the bidder reported any.
+		if len(adapterExtra[a].Errors) > 0 {
+			bidResponseExt.Errors[a] = adapterExtra[a].Errors
+		}
+		if len(errList) > 0 {
+			s := make([]string, len(errList))
+			for i :=0; i<len(errList); i++ {
+				s[i] = errList[i].Error()
+			}
+			bidResponseExt.Errors["prebid"] = s
+		}
+		bidResponseExt.ResponseTimeMillis[a] = adapterExtra[a].ResponseTimeMillis
+		// Defering the filling of bidResponseExt.Usersync[a] until later
+
+	}
+	return bidResponseExt
+}
+
+// Return an openrtb seatBid for a bidder
+// BuildBidResponse is responsible for ensuring nil bid seatbids are not included
+func (e *exchange) MakeSeatBid(adapterBid *pbsOrtbSeatBid, adapter openrtb_ext.BidderName, adapterExtra map[openrtb_ext.BidderName]*seatResponseExtra) *openrtb.SeatBid {
+	seatBid := new(openrtb.SeatBid)
+	seatBid.Seat = adapter.String()
+	// Prebid cannot support roadblocking
+	seatBid.Group = 0
+
+	sbExt := ExtSeatBid{
+		Bidder: adapterBid.ext,
+	}
+
+	ext, err := json.Marshal(sbExt)
+	if err != nil {
+		adapterExtra[adapter].Errors = append(adapterExtra[adapter].Errors, fmt.Sprintf("Error writing SeatBid.Ext: %s", err.Error()))
+	}
+	seatBid.Ext = ext
+	var errList []string
+	seatBid.Bid, errList = e.MakeBid(adapterBid.bids)
+	if len(errList) > 0 {
+		adapterExtra[adapter].Errors = append(adapterExtra[adapter].Errors, errList...)
+	}
+
+	return seatBid
+}
+
+// Create the Bid array inside of SeatBid
+func (e *exchange) MakeBid(Bids []*pbsOrtbBid) ([]openrtb.Bid, []string) {
```

Comment:
The overhead of creating an `exchange` and `adapterMap` is not too bad since I already deployed code elsewhere in the test suite to create these objects. makeBids takes a lot more parameters when I need to start supporting targeting, so I don't know I want to change too much in here until targeting is merged in.

Your suggestion for a disposable auction above might be a better way of handling these things given the parameter explosion needed to support targeting.